include/swift/Remote/MetadataReader.h

```diff
@@ -1435,27 +1459,28 @@ class MetadataReader {
       return nullptr;
     }
 
-    // Override the node kind if this was a Clang-imported type.
-    if (isTypeContext) {
-      auto typeFlags =
-        TypeContextDescriptorFlags(descriptor->Flags.getKindSpecificFlags());
-
-      if (typeFlags.isCTypedef())
+    // Override various aspects of the mangling for imported types.
+    if (importInfo && isCImportedContext(parentDemangling)) {
+      if (importInfo->SymbolNamespace == TypeImportSymbolNamespace::CTypedef)
         nodeKind = Demangle::Node::Kind::TypeAlias;
 
-      // As a special case, always use the struct mangling for C-imported
-      // value types.
-      else if (nodeKind == Demangle::Node::Kind::Enum &&
-               !typeFlags.isSynthesizedRelatedEntity() &&
-               isCImportedContext(parentDemangling))
+      // As a special case, use the struct mangling for C-imported value
+      // types that don't have a special namespace and aren't a related
+      // entity.
+      //
+      // This should be kept in sync with _isCImportedTagType in the runtime.
```

Comment:
This should really be kept in sync with the mangler itself. That's the point of truth for this thing.